File: 1477-find-two-non-overlapping-sub-arrays-each-with-target-sum/1477-find-two-non-overlapping-sub-arrays-each-with-target-sum.c
#include <stdlib.h>

#define MIN(a, b) ((a) < (b) ? (a) : (b))

int minSumOfLengths(int* arr, int arrSize, int target) {
    int n = arrSize;
    int INF = n + 1;

    int* best = (int*)malloc((n + 1) * sizeof(int));
    for (int i = 0; i <= n; i++) {
        best[i] = INF;
    }

    int left = 0;
    int sum = 0;
    int shortest = INF;
    int answer = INF;

    for (int right = 0; right < n; right++) {
        sum += arr[right];

        while (sum > target) {
            sum -= arr[left++];
        }

        if (sum == target) {
            int length = right - left + 1;

            if (best[left] != INF) {
                answer = MIN(answer, best[left] + length);
            }
            shortest = MIN(shortest, length);
        }
        best[right + 1] = shortest;
    }

    free(best);
    return answer == INF ? -1 : answer;
}

File: 1477-find-two-non-overlapping-sub-arrays-each-with-target-sum/1477-find-two-non-overlapping-sub-arrays-each-with-target-sum.py
class Solution:
    def minSumOfLengths(self, arr: list[int], target: int) -> int:
        n = len(arr)
        INF = n + 1

        # Correct bracket list initialization to avoid syntax bugs
        best = [INF] * (n + 1)

        left = 0
        sum_val = 0
        shortest = INF
        answer = INF

        for right in range(n):
            sum_val += arr[right]

            while sum_val > target:
                sum_val -= arr[left]
                left += 1

            if sum_val == target:
                length = right - left + 1

                if best[left] != INF:
                    answer = min(answer, best[left] + length)
                shortest = min(shortest, length)
                
            best[right + 1] = shortest

        return -1 if answer == INF else answer
